### PR TITLE
Revert "Revert "[Enhancement] Optimize hash join when hash table is empty" (#9701)"

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [x] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9701  #901

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/exec/vectorized/join_hash_map.h [hunk 1]
```diff
@@ -434,6 +435,98 @@ class SerializedJoinProbeFunc {
...
+
+private:
+    void _probe_output(ChunkPtr* probe_chunk, ChunkPtr* chunk) {
+        bool to_nullable = _table_items->left_to_nullable;
+        for (size_t i = 0; i < _table_items->probe_column_count; i++) {
+            HashTableSlotDescriptor hash_table_slot = _table_items->probe_slots[i];
+            SlotDescriptor* slot = hash_table_slot.slot;
+            auto& column = (*probe_chunk)->get_column_by_slot_id(slot->id());
+            if (hash_table_slot.need_output) {
+                if (to_nullable && !column->is_nullable()) {
```
Comment (LiShuMing) [inline be/src/exec/vectorized/join_hash_map.h:490]: here should take care column's nullable.

**Review by satanson — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by LiShuMing** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Comment by LiShuMing** [comment]
@mergify branch-2.4

**Comment by LiShuMing** [comment]
@mergify backport branch-2.4

**Pull request merged**